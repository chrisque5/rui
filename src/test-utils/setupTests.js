// The below part of the code is to fix the issue
// TypeError: window.matchMedia is not a function
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});
