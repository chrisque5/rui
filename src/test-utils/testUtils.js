/* eslint-disable import/no-extraneous-dependencies, react/prop-types */
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Form } from 'antd';

const middlewares = [thunk];
const renderWithForm = (
  ui,
  {
    initialState,
    store = configureMockStore(middlewares)(initialState),
    ...renderOptions
  } = {},
) => {
  const wrapper = ({ children }) => <Provider store={store}><Form>{children}</Form></Provider>;

  return render(ui, { wrapper, ...renderOptions });
};

const renderWithProvider = (
  ui,
  {
    initialState,
    store = configureMockStore(middlewares)(initialState),
    ...renderOptions
  } = {},
) => {
  const wrapper = ({ children }) => <Provider store={store}>{children}</Provider>;

  return render(ui, { wrapper, ...renderOptions });
};
// re-export everything
export * from '@testing-library/react';
// override render method
export { renderWithForm, renderWithProvider };
