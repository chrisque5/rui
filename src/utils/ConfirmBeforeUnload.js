import { useEffect, useCallback } from 'react';

const ConfirmBeforeUnload = () => {
  const onBeforeUnload = useCallback((e) => {
    e.preventDefault();
    e.returnValue = '';
  }, []);

  useEffect(() => {
    window.addEventListener('beforeunload', onBeforeUnload);

    return () => {
      window.removeEventListener('beforeunload', onBeforeUnload);
    };
  }, [onBeforeUnload]);

  return null;
};

export default ConfirmBeforeUnload;
