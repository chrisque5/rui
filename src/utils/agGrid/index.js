import {
  approvalsCellRenderer, approvalsSummaryCellRenderer, isUnderInvestigationCellRenderer, stpCellRenderer, stpFilterCellRenderer,
} from './js-renderers';
import {
  dateFormatter, formatDate, numberFormatter, timeFormatter,
  valueDateFormatter, fixingDateFormatter, publishDateFormatter,
  priceFormatter, brokerageFormatter,
} from './formatters';
import { dateFilterComparator, numberFilterComparitor, timeFilterComparator } from './comparitors';
import { getExportMenuItems } from './exportUtils';

export {
  approvalsCellRenderer,
  approvalsSummaryCellRenderer,
  dateFilterComparator,
  dateFormatter,
  formatDate,
  getExportMenuItems,
  isUnderInvestigationCellRenderer,
  numberFilterComparitor,
  numberFormatter,
  stpCellRenderer,
  stpFilterCellRenderer,
  timeFilterComparator,
  timeFormatter,
  valueDateFormatter,
  fixingDateFormatter,
  publishDateFormatter,
  priceFormatter,
  brokerageFormatter,
};
