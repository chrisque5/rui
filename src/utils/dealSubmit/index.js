import mapFormToFwdChain from './mapFormToFwdChain';
import mapFormToNdfChain from './mapFormToNdfChain';
import mapFormToSptChain from './mapFormToSptChain';

export { mapFormToFwdChain, mapFormToNdfChain, mapFormToSptChain };
