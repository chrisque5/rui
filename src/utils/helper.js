import moment from 'moment';
import * as math from 'mathjs';
import { Modal } from 'antd';

export const isValidTerm = (termToTest, selections) => {
  if (termToTest.length < 2) return false;

  const lowerCaseTermToTest = termToTest.toLowerCase();

  const endsWithPeriod = (term) => {
    const periods = ['d', 'w', 'm', 'y'];
    return (periods.includes(term.charAt(term.length - 1)));
  };

  const firstCharacterIsNumber = (term) => {
    const result = term.substring(0, term.length - 1).match(/^[0-9]*$/);
    return result ? result.length > 0 : false;
  };

  const isStaticSelectionValid = (term) => {
    const validStaticTerms = selections.map((x) => x.toLowerCase());
    if (validStaticTerms.includes(term)) {
      return true;
    }

    if (term.startsWith('bmf') || term.startsWith('imm')) {
      let termDaysOnly = term.replace('bmf', '');
      termDaysOnly = termDaysOnly.replace('imm', '');
      const beforeConversion = termDaysOnly;
      termDaysOnly = parseInt(termDaysOnly, 10);
      if (beforeConversion.length === termDaysOnly.toString().length) {
        return (termDaysOnly % 1 === 0 && termDaysOnly > 0 && termDaysOnly < 101);
      }
    }
    return false;
  };

  return isStaticSelectionValid(lowerCaseTermToTest)
    || (endsWithPeriod(lowerCaseTermToTest) && firstCharacterIsNumber(lowerCaseTermToTest));
};

export const doesArrayExistWithValue = (array, index = 0) => {
  if (Array.isArray(array)) {
    if (array && array.length > 0 && array[index]) {
      return true;
    }
  }
  return false;
};

export const setTodayOnNull = (form, id, value) => {
  if (!value) {
    setTimeout(() => {
      form.setFieldsValue({ [id]: moment() });
    });
  }
};

export const dayDifferenceIgnoreHours = (earlierDateMoment, laterDateMoment) => {
  const date1 = earlierDateMoment.clone().startOf('day');
  const date2 = laterDateMoment.clone().startOf('day');
  const dayCount = date1.diff(date2, 'days');

  return dayCount;
};

export const getDateInLocalTZ = () => moment().tz(moment.tz.guess(true)).format('YYYY-MM-DD');
export const getDateTimeInLocalTZ = () => moment().tz(moment.tz.guess(true));

export const parseValueAsBigNumber = (value) => {
  try {
    return math.bignumber(value);
  } catch (err) {
    return null;
  }
};

export const formatAmount = (val) => {
  const inputValue = val.replace(/\s/g, '');
  let multiplier = 1;
  const oneM = 1000000;
  const oneB = 1000000000;
  let amt = inputValue;

  if (Number.isNaN(Number(inputValue)) && inputValue.length > 1) {
    const lastChar = inputValue.charAt(inputValue.length - 1).toLowerCase();
    amt = Number(inputValue.substring(0, inputValue.length - 1));
    if (lastChar === 'm') {
      multiplier = oneM;
    } else if (lastChar === 'b') {
      multiplier = oneB;
    } else {
      return inputValue;
    }
  } else if (Number(inputValue) < oneM) {
    multiplier = oneM;
  }

  const calculatedAmount = math.multiply(math.bignumber(amt), math.bignumber(multiplier));
  return calculatedAmount.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
};

export const isValidDate = (date, dateFormat) => moment(date, dateFormat || 'L', true).isValid();

export const findCurrencyPair = (ccy1, ccy2, currencies) => (
  currencies.find(
    (currency) => currency.baseCurrency === ccy1 && currency.counterCurrency === ccy2,
  )
);

export const filterSpotExecutionVenues = (execVenues) => execVenues.filter((i) => i.venueId === 'XOFF');

export const isDealTypeSame = (dealType1, dealType2) => (dealType1 === dealType2);

export const isStrategyTypeSame = (strategyType1, strategyType2) => (strategyType1 === strategyType2);

export const ConfirmUnsavedChanges = (message, onYes, onNo) => {
  Modal.confirm({
    title: 'Confirm',
    content: message,
    okText: 'Yes',
    cancelText: 'No',
    onOk: () => {
      if (onYes) {
        setTimeout(() => {
          onYes();
        });
      }
    },
    onCancel: () => {
      if (onNo) {
        setTimeout(() => {
          onNo();
        });
      }
    },
  });
};

export const isObjectEmpty = (obj) => Object.keys(obj).length === 0 && obj.constructor === Object;

export const getUniqueID = () => {
  const s4 = () => Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
  return `${s4() + s4()}-${s4()}`;
};

export const showPreviousValueToolTip = (currentValue, previousValue) => {
  if (currentValue !== previousValue) {
    return { title: `Previous Value: ${previousValue || 'N/A'}`, className: 'deal-input-dirty' };
  }
  return { title: '', className: '' };
};
