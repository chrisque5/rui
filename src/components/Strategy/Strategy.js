import PropTypes from 'prop-types';
import { Form, Radio } from 'antd';
import { useDispatch } from 'react-redux';
import { ids, actionTypes } from '../../utils/constants';

const Strategy = ({ options, onChangeAfterEffects }) => {
  const dispatch = useDispatch();

  const strategyChanged = (event) => {
    dispatch({ type: actionTypes.CHANGE_STRATEGY, strategy: event.target.value });
    onChangeAfterEffects();
  };

  return (
    <Form.Item
      name={ids.STRATEGY}
      rules={[{ required: true, message: 'Strategy required' }]}
    >
      <Radio.Group>
        {options.map((strategy) => (
          <Radio
            key={strategy.name}
            value={strategy.name}
            onChange={strategyChanged}
            disabled={strategy.isDisabled}
            data-testid={`rdoStrategy${strategy.name}`}
          >
            {strategy.name}
          </Radio>
        ))}
      </Radio.Group>
    </Form.Item>
  );
};

Strategy.defaultProps = {
  onChangeAfterEffects: () => {},
};

Strategy.propTypes = {
  options: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  onChangeAfterEffects: PropTypes.func,
};

export default Strategy;
