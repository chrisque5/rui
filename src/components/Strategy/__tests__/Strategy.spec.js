import { render, waitFor, fireEvent } from 'test-utils/testUtils';
import redux from 'react-redux';
import { Form } from 'antd';
import Strategy from '../Strategy';
import { strategyOptions } from '../../../utils/constants';

jest.mock('react-redux', () => ({
  useDispatch: () => () => { },
}));

const form = {};
const dispatchSpy = jest.spyOn(redux, 'useDispatch');

const options = strategyOptions.NDF;
const onChangeAfterEffects = jest.fn();

const defaultProps = {
  form,
  options,
  onChangeAfterEffects,
};

describe('<Strategy />', () => {
  test('It Renders ', () => {
    const { container } = render(<Form><Strategy {...defaultProps} /></Form>);
    expect(container).toMatchSnapshot();
  });

  test('Switch to NDF Spread', async () => {
    const { getByTestId } = render(<Form><Strategy {...defaultProps} /></Form>);
    const rdoStrategySpread = await waitFor(() => getByTestId('rdoStrategySpread'));
    fireEvent.click(rdoStrategySpread);
    expect(dispatchSpy).toHaveBeenCalled();
    expect(onChangeAfterEffects).toHaveBeenCalled();
  });

  test('Switch to NDF Outright', async () => {
    const { getByTestId } = render(<Form><Strategy {...defaultProps} /></Form>);
    const rdoStrategyOutright = await waitFor(() => getByTestId('rdoStrategyOutright'));
    fireEvent.click(rdoStrategyOutright);
    expect(dispatchSpy).toHaveBeenCalled();
    expect(onChangeAfterEffects).toHaveBeenCalled();
  });

  test('Switch to FWD Forward', async () => {
    const { getByTestId } = render(<Form><Strategy {...defaultProps} options={strategyOptions.FWD} /></Form>);
    const rdoStrategyForward = await waitFor(() => getByTestId('rdoStrategyForward'));
    fireEvent.click(rdoStrategyForward);
    expect(dispatchSpy).toHaveBeenCalled();
    expect(onChangeAfterEffects).toHaveBeenCalled();
  });

  test('Switch to FWD Forward Forward', async () => {
    const { getByTestId } = render(<Form><Strategy {...defaultProps} options={strategyOptions.FWD} /></Form>);
    const rdoStrategyForward = await waitFor(() => getByTestId('rdoStrategyForward Forward'));
    fireEvent.click(rdoStrategyForward);
    expect(dispatchSpy).toHaveBeenCalled();
    expect(onChangeAfterEffects).toHaveBeenCalled();
  });

  test('Switch to FWD Outright', async () => {
    const { getByTestId } = render(<Form><Strategy {...defaultProps} options={strategyOptions.FWD} /></Form>);
    const rdoStrategyOutright = await waitFor(() => getByTestId('rdoStrategyOutright'));
    fireEvent.click(rdoStrategyOutright);
    expect(dispatchSpy).toHaveBeenCalled();
    expect(onChangeAfterEffects).toHaveBeenCalled();
  });
});
