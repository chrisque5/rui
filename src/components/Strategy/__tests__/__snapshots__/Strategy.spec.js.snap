// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Strategy /> It Renders  1`] = `
<div>
  <form
    class="ant-form ant-form-horizontal"
  >
    <div
      class="ant-row ant-form-item"
    >
      <div
        class="ant-col ant-form-item-control"
      >
        <div
          class="ant-form-item-control-input"
        >
          <div
            class="ant-form-item-control-input-content"
          >
            <div
              class="ant-radio-group ant-radio-group-outline"
              id="strategy"
            >
              <label
                class="ant-radio-wrapper"
              >
                <span
                  class="ant-radio"
                >
                  <input
                    class="ant-radio-input"
                    data-testid="rdoStrategyOutright"
                    type="radio"
                    value="Outright"
                  />
                  <span
                    class="ant-radio-inner"
                  />
                </span>
                <span>
                  Outright
                </span>
              </label>
              <label
                class="ant-radio-wrapper"
              >
                <span
                  class="ant-radio"
                >
                  <input
                    class="ant-radio-input"
                    data-testid="rdoStrategySpread"
                    type="radio"
                    value="Spread"
                  />
                  <span
                    class="ant-radio-inner"
                  />
                </span>
                <span>
                  Spread
                </span>
              </label>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
